api/internal/crawl/cmd/kustomize_stats/main.go

```diff
@@ -0,0 +1,141 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"sigs.k8s.io/kustomize/api/internal/crawl/index"
+	"sigs.k8s.io/kustomize/api/konfig"
+)
+
+// iterateArr adds each item in arr into countMap.
+func iterateArr(arr []string, countMap map[string]int) {
+	for _, item := range arr {
+		if _, ok := countMap[item]; !ok {
+			countMap[item] = 1
+		} else {
+			countMap[item]++
+		}
+	}
+
+}
+
+// isKustomizationFile determines whether a file path is a kustomization file
+func isKustomizationFile(path string) bool {
+	basename := filepath.Base(path)
+	for _, name := range konfig.RecognizedKustomizationFileNames() {
+		if basename == name {
+			return true
+		}
+	}
+	return false
+}
+
+func main() {
+	topKindsPtr := flag.Int("kinds", -1,
+		"the number of kubernetes object kinds to be listed according to their popularities." +
+		"(By default, all the kinds will be listed. " +
+		"If you only want to list the 10 most popular kinds, set the flag to 10.)")
+	topIdentifiersPtr := flag.Int("identifiers", -1,
+		"the number of identifiers to be listed according to their popularities." +
+		"(By default, all the identifiers will be listed. " +
+		"If you only want to list the 10 most popular identifiers, set the flag to 10.)")
+	topKustomizeFeaturesPtr := flag.Int("kustomize-features", -1,
+		"the number of kustomize features to be listed according to their popularities." +
+			"(By default, all the features will be listed. " +
+			"If you only want to list the 10 most popular features, set the flag to 10.)")
+	flag.Parse()
+
+	ctx := context.Background()
+	idx, err := index.NewKustomizeIndex(ctx)
+	if err != nil {
+		fmt.Printf("Could not create an index: %v\n", err)
+		return
+	}
+
+	// count tracks the number of documents in the index
+	count := 0
+
+	// kustomizationFilecount tracks the number of kustomization files in the index
+	kustomizationFilecount := 0
+
+	kindsMap := make(map[string]int)
+	identifiersMap := make(map[string]int)
+	kustomizeIdentifiersMap := make(map[string]int)
+
+	// ids tracks the unique IDs of the documents in the index
+	ids := make(map[string]struct{})
+
+	// get all the documents in the index
+	query := []byte(`{ "query":{ "match_all":{} } }`)
+	it := idx.IterateQuery(query, 10000, 60*time.Second)
+	for it.Next() {
+		for _, hit := range it.Value().Hits.Hits {
+			// check whether there is any duplicate IDs in the index
+			if _, ok := ids[hit.ID]; !ok {
+				ids[hit.ID] = struct{}{}
+			} else {
+				fmt.Printf("Found duplicate ID (%s)\n", hit.ID)
+			}
+
+			count++
+			iterateArr(hit.Document.Kinds, kindsMap)
+			iterateArr(hit.Document.Identifiers, identifiersMap)
+
+			if isKustomizationFile(hit.Document.FilePath) {
+				kustomizationFilecount++
+				iterateArr(hit.Document.Identifiers, kustomizeIdentifiersMap)
+			}
+		}
+	}
+	if err := it.Err(); err != nil {
+		fmt.Printf("Error iterating: %v\n", err)
+	}
+
+	sortMapKeyByValueFunc := func(m map[string]int) []string {
```

Comment:
please just declare this outside main like `isKustomizationFile`
(main's a bit too long for my brain)

Reply:
Done